aml/src/stream.rs

```diff
@@ -0,0 +1,339 @@
+use core::{convert::TryInto, mem};
+
+/// A container for the buffer that provides debugging info
+#[derive(Copy, Clone)]
+pub struct AmlStream<'a> {
+    /// The address of the original data, used for debugging only
+    stream_start: usize,
+    /// The slice of data we are currently processing
+    buf: &'a [u8],
+}
+
+/// Slice operations for the stream - if there is an error, the returned stream will be empty
+impl<'a> AmlStream<'a> {
+    pub fn from_slice(slice: &'a [u8]) -> Self {
+        Self { stream_start: slice.as_ptr() as usize, buf: slice }
+    }
+
+    pub fn empty() -> Self {
+        Self { stream_start: 0, buf: &[] }
+    }
+
+    pub fn len(&self) -> usize {
+        self.buf.len()
+    }
+
+    pub fn data(&self) -> &[u8] {
+        self.buf
+    }
+
+    pub fn take(&self, start: usize, end: usize) -> Self {
+        if start > self.len() || end > self.len() || start > end {
+            return Self::empty();
+        }
+        Self { stream_start: self.stream_start, buf: &self.buf[start..end] }
```

Comment:
Hm, I thought we'd be advancing this as we made our way through the stream? Maybe we're talking at cross purposes here - is the intention just to track the entire stream along with the part we're parsing? I thought we were trying to add context to the slice in terms of an offset into the whole slice etc.?